Deprecate buildNeeded and buildDependents tasks (#37574)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_9.adoc

implementation.getDependencies().add(dependency);
----

[[deprecate_build_needed_build_dependents_tasks]]
==== Deprecation of `buildNeeded` and `buildDependents` tasks

> The `buildNeeded` and `buildDependents` tasks registered by the Java plugins have been deprecated and will be removed in Gradle 10.

Reviewer: ```suggestion
The `buildNeeded` and `buildDependents` tasks from the Java plugins are deprecated and will be removed in Gradle 10.
```

> `buildNeeded` intends to build all projects that the current project depends on, while `buildDependents` intends to build all projects that depend on the current project.

Reviewer: ```suggestion
`buildNeeded` builds the current project along with all projects it depends on.
`buildDependents` does the opposite: it builds the current project and all projects that depend on it.
```

> These tasks were introduced back in Gradle 0.8, and while they serve valid use cases, they are particularly inflexible and are implemented in a manner incompatible with Configure on Demand and Isolated Projects.

Reviewer: ```suggestion
These tasks were introduced back in Gradle 0.8.
Their implementation is incompatible with Configure on Demand and Isolated Projects, and the tasks themselves are too rigid to cover the range of use cases users actually need.
```

> Rather than modernize these tasks, we have decided to deprecate and remove them in favor of existing Gradle functionality.

Reviewer: ```suggestion
Existing Gradle functionality covers these use cases, so rather than rework the tasks, we are removing them.
```

> `buildNeeded` depends on the `build` task of the current project, and all `buildNeeded` tasks of all projects that are direct dependencies of the `testRuntimeClasspath` configuration.

Reviewer: ```suggestion
`buildNeeded` depends on the current project's `build` task and on the `buildNeeded` task of each project that appears as a direct dependency in the `testRuntimeClasspath` configuration.
Because those `buildNeeded` tasks depend on each other, the result is that the entire project dependency closure is built.
```

> The following options may replace existing usages of this task:

Reviewer: ```suggestion
Depending on what you were using `buildNeeded` for, one of the following may replace it:
```

> - Run the specific tasks you would like to execute (e.g., `./gradlew :proj:test` or `./gradlew :proj:build`), and rely on task dependency resolution to automatically execute required task dependencies.

Reviewer: ```suggestion
- If you only need to build or test a specific project, run that task directly (e.g., `./gradlew :proj:build`) and let Gradle resolve its task dependencies automatically.
```

> - Use <<test_report_aggregation_plugin.adoc#test_report_aggregation_plugin, test report aggregation>> to execute tests across all projects that the current project depends on.

Reviewer: ```suggestion
- For running tests across all projects the current project depends on, use <<test_report_aggregation_plugin.adoc#test_report_aggregation_plugin, test report aggregation>>.
```

> - Use <<artifact_views.adoc#artifact-views, Artifact Views>> to select and depend on work associated with all projects in a particular dependency graph.

Reviewer: ```suggestion
- For plugin authors and advanced build logic, use <<artifact_views.adoc#artifact-views, Artifact Views>> to select artifacts from projects in a dependency graph and wire their producer tasks as dependencies.
```

> `buildDependents` depends on the `build` task of the current project, then configures all projects in the build and depends on all `buildDependents` tasks in all projects where the `testRuntimeClasspath` configuration in that project depends on the current project.

Reviewer: We already said this up above, not sure if you want to say it again. I would either delete this or update it but i'm leaning towards **delete**:
```suggestion
`buildDependents` depends on the current project's `build` task. 
It then configures every project in the build and depends on the `buildDependents` task of each project whose `testRuntimeClasspath` includes the current project.
```

> `buildDependents` depends on the `build` task of the current project, then configures all projects in the build and depends on all `buildDependents` tasks in all projects where the `testRuntimeClasspath` configuration in that project depends on the current project.

Reviewer: I would prefer to delete this since we said it already, and it breaks the flow:
```suggestion
```

> To replace `buildDependents`, instead execute the desired tasks directly (e.g., `./gradlew test` or `./gradlew build`) and rely on up-to-date checks and build caching to avoid executing work not affected by current changes.

Reviewer: ```suggestion
There is no direct replacement for `buildDependents`. 
For most workflows, running the desired task from the root (e.g., `./gradlew test`) combined with up-to-date checks and build caching will skip work for unaffected projects, at the cost of configuring all projects. 
```

[[deprecate_getTaskDependencyFromProjectDependency]]
==== Deprecation of `Configuration.getTaskDependencyFromProjectDependency()`

The link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#getTaskDependencyFromProjectDependency(boolean,java.lang.String)[`Configuration.getTaskDependencyFromProjectDependency(boolean, String)`] method has been deprecated and will be removed in Gradle 10.

> This method is used to implement the above deprecated `buildNeeded` and `buildDependents` tasks.

Reviewer: ```suggestion
It exists to support the `buildNeeded` and `buildDependents` tasks deprecated above, and has no other intended use.
```

> The implementation of this method is fundamentally incompatible with Isolated Projects and Configure on Demand.

Reviewer: ```suggestion
Its implementation is fundamentally incompatible with Isolated Projects and Configure on Demand.

```

There is no direct replacement for this method.

> 

Reviewer: ```suggestion
```

> To run tasks across different projects in a build, prefer <<artifact_views.adoc#artifact-views, Artifact Views>> to select particular artifacts from projects that the current project depends on.

Reviewer: ```suggestion
If you were using it to reach tasks across project boundaries, <<artifact_views.adoc#artifact-views, Artifact Views>> let you select artifacts from dependent projects and let Gradle wire the producer tasks through normal dependency resolution.
```
